pyembed/src/importer.rs

```diff
@@ -900,7 +904,219 @@ impl OxidizedFinder {
+    }
+
+    // Canonicalize the path to the current executable or raise an OSError.
+    //
+    // Even `std::env::current_exe()` can fail to `fs::canonicalize()` if, e.g.,
+    // the file is deleted after the program starts.
+    fn exe_realpath(&self, py: Python) -> PyResult<std::path::PathBuf> {
+        let current_exe = &self.state(py).get_resources_state().current_exe;
+        current_exe.canonicalize().map_err(|err| {
+            let exe = current_exe.display();
+            // Use a Python expression instead of PyErr::new to ensure we get the
+            // right OSError subclass.
+            let strerror = format!("cannot open current executable: '{}'", exe);
+            let raw_os_error = err
+                .raw_os_error()
+                .map_or_else(|| "None".to_string(), |err_code| err_code.to_string());
+            let code = if cfg!(windows) {
+                format!(
+                    "OSError(None, r\"{}\", r\"{}\", {})",
+                    strerror, exe, raw_os_error
+                )
+            } else {
+                format!("OSError({}, r\"{}\", r\"{}\")", raw_os_error, strerror, exe)
+            };
+            py.eval(&code, None, None).map_or_else(
+                |err_creating_exc| err_creating_exc,
+                |exc| PyErr::from_instance(py, exc),
+            )
+        })
+    }
+
+    fn path_hook_impl(&self, py: Python, path: PyObject) -> PyResult<_PathEntryFinder> {
+        // Compute _PathEntryFinder::package
+        let pkg = {
+            let current_exe = self.exe_realpath(py)?;
+            let not_exe_err = || {
+                let msg = format!(
+                    "path {} does not begin in \'{}\' (if you're sure it does, check Python's file-system encoding)", &path, current_exe.display());
+                let mut err = PyErr::new::<ImportError, _>(py, msg);
+                let exc = err.instance(py);
+                if let Err(setattr_err) = exc.setattr(py, "path", path.clone_ref(py)) {
+                    setattr_err
+                } else {
+                    err
+                }
+            };
+            let pbuf = pyobject_to_pathbuf(py, path.clone_ref(py))?;
+            let mut p: &std::path::Path = pbuf.as_ref();
+            // path could point "inside" current_exe. If we can't canonicalize,
+            // strip the last component and check again. (zipimporter does
+            // somethings similar)
+            let mut abs_p = p.canonicalize();
+            // Count the number of components at the end of `path` will be in pkg
+            let mut parts: usize = 0;
+            while abs_p.is_err() {
+                p = match p.parent() {
+                    Some(parent) => {
+                        parts += 1;
+                        parent
+                    }
+                    None => {
+                        return Err(not_exe_err());
+                    }
+                };
+                abs_p = p.canonicalize();
+            }
```

Comment:
We can likely use `Path.components()` plus iterators here to do something more elegantly. Although this code isn't horrible. Another strategy would be to use https://doc.rust-lang.org/std/path/struct.Path.html#method.strip_prefix to remove the exe part of the path. e.g. `abs_p.strip_prefix(current_exe).map_err(...).components().iter().count()` (or something like that - we may need to filter the components as part of the iterator bit).

Reply:
I had considered [`Path::strip_prefix`] here and below, but here's a bug that could arise from that. Suppose `current_exe` is `/path/to/myapp` but `sys.executable` is `/symlink/to/myapp` and that both are correct (`sys.executable` often points to a symlink when Python is running inside a virtual environment).
```python
pkgutil.iter_modules([os.path.join(sys.executable, 'importlib')])
```
should list all the modules in the `importlib` package, but will instead be empty because `/symlink/to/myapp/importlib` does not start with `/path/to/myapp`. (Maybe we only care that `pkgutil.iter_modules(importlib.__path__)` works, which it would without canonicalization. But that seems to me to be a confusing API.) Calling `canonicalize` on `/symlink/to/myapp/importlib` will fail because that path doesn't exist. So we either need all the paths that point at `/path/to/myapp` (impossible) and use  [`Path::strip_prefix`] until one works (slow), or we need to peel the end of the path off one component at a time like I do here or like [`zipimport` does](https://github.com/python/cpython/blob/b36349a647b2bf8174f0e736a4fc347e92ae204e/Lib/zipimport.py#L74-L90).

I am now realizing I don't have this as a test case, so I'll add it.

[`Path::strip_prefix`]: https://doc.rust-lang.org/std/path/struct.Path.html#method.strip_prefix

Reply:
I added the symlink test case in 9cbaf7c3a38817eda0e9a1e88cb4c598d5e2051a but I haven't run it on Windows yet where I used a directory junction since symlinks are security cordoned there